### PR TITLE
Add cowsrevenge game

### PR DESCRIPTION
**Pull request by Dariasteam**

### CODE REVIEW

**Comment by nedrichards** [comment]
bot, build org.pipoypipagames.cowsrevenge

File: org.pipoypipagames.cowsrevenge.json [hunk 1]
```diff
@@ -0,0 +1,29 @@
+{
+  "app-id": "org.pipoypipagames.cowsrevenge",
+  "branch": "stable",
+  "runtime": "org.freedesktop.Platform",
+  "runtime-version": "1.6",
+  "sdk": "org.freedesktop.Sdk",
+  "command": "cowsrevenge",
+  "finish-args": [
+     "--socket=x11",
```
Comment (nedrichards) [inline org.pipoypipagames.cowsrevenge.json]: This also needs --share=ipc in order to get x11 to work at top performance. http://docs.flatpak.org/en/latest/sandbox-permissions.html

File: org.pipoypipagames.cowsrevenge.json [hunk 2]
```diff
@@ -0,0 +1,29 @@
+{
+  "app-id": "org.pipoypipagames.cowsrevenge",
```
Comment (nedrichards) [inline org.pipoypipagames.cowsrevenge.json]: Do you own this domain name? It looks like it's available. You should control the path of your app-id. e.g. you could use com.github.dariasteam.cowsrevenge

File: org.pipoypipagames.cowsrevenge.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  "branch": "stable",
+  "runtime": "org.freedesktop.Platform",
+  "runtime-version": "1.6",
+  "sdk": "org.freedesktop.Sdk",
+  "command": "cowsrevenge",
+  "finish-args": [
+     "--socket=x11",
+     "--socket=pulseaudio",
+     "--filesystem=home"
```
[the remark this answers is not in the document]
Comment (Dariasteam) [reply org.pipoypipagames.cowsrevenge.json]: for save files, initially I thought allowing only ~/.godot folder (the only which actually is used) but didn't know if I should manage its creation or let godot itself do the work as it is supossed to be

File: org.pipoypipagames.cowsrevenge.json (hunk 1 above)
Comment (Dariasteam) [reply org.pipoypipagames.cowsrevenge.json]: Didn't know that, thanks

File: org.pipoypipagames.cowsrevenge.json (hunk 2 above)
Comment (Dariasteam) [reply org.pipoypipagames.cowsrevenge.json]: okay, it was certainly completely arbitrary

File: org.pipoypipagames.cowsrevenge.json [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    "split" : false,
+    "no-debuginfo" : true
+  },
+  "modules": [
+    {
+      "name": "cowsrevenge",
+      "sources": [
+        {
+          "type": "git",
+          "url": "https://github.com/Dariasteam/Cows-Revenge-Flatpak",
```
Comment (nedrichards) [inline org.pipoypipagames.cowsrevenge.json]: Which architectures does this work on? It looks like there's just an x86_64 binary at this domain. You need to use a flathub.json file to prevent it being offered on other systems where it won't work.

Example here: https://github.com/flathub/us.zoom.Zoom/commit/c669b61fa49c032b8c2220f8622af33effaec390

File: org.pipoypipagames.cowsrevenge.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (nedrichards) [reply org.pipoypipagames.cowsrevenge.json]: Have a look at the sandbox permissions documentation. http://docs.flatpak.org/en/latest/sandbox-permissions.html you want "--persist=.godot", instead

File: org.pipoypipagames.cowsrevenge.json (hunk 4 above)
Comment (Dariasteam) [reply org.pipoypipagames.cowsrevenge.json]: Thank you, gonna change it now

File: org.pipoypipagames.cowsrevenge.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+          "sha256" : "cc84432ba9b9d1b85b5f1484a5930fe5e5f02b97314a06714cfa69d07e290ea0"
```
Comment (nedrichards) [inline org.pipoypipagames.cowsrevenge.json]: This should be a commit rather than a sha

File: org.pipoypipagames.cowsrevenge.json (hunk 4 above)
Comment (Dariasteam) [reply org.pipoypipagames.cowsrevenge.json]: There is some list with all architectures codenames? Can't find anything in flatpak documentation

Comment (nedrichards) [reply org.pipoypipagames.cowsrevenge.json]: "aarch64", "arm" and "i386" are the ones you want.

**Comment by nedrichards** [comment]
bot, build com.github.dariasteam.cowsrevenge

File: com.github.dariasteam.cowsrevenge.json [hunk 6]
```diff
@@ -0,0 +1,31 @@
+{
+  "app-id": "com.github.dariasteam.cowsrevenge",
+  "branch": "stable",
+  "runtime": "org.freedesktop.Platform",
+  "runtime-version": "1.6",
+  "sdk": "org.freedesktop.Sdk",
+  "skip-arches": ["aarch64", "arm"],
```
Comment (nedrichards) [inline com.github.dariasteam.cowsrevenge.json]: ah! sorry, you need this bit in a separate file called flathub.json. This is a flathub only file rather than a generic thing for flatpak-builder.

**Comment by nedrichards** [comment]
bot, build com.github.dariasteam.cowsrevenge

**Comment by nedrichards** [comment]
Build fails https://flathub.org/builds/#/builders/28/builds/93 you need to update the app id here as well https://github.com/Dariasteam/Cows-Revenge-Flatpak/blob/master/com.github.dariasteam.cowsrevenge.xml#L14

**Comment by Dariasteam** [comment]
Ouch sorry! Should be fixed now

File: com.github.dariasteam.cowsrevenge.json [hunk 7]
```diff
@@ -0,0 +1,30 @@
+{
+  "app-id": "com.github.dariasteam.cowsrevenge",
+  "branch": "stable",
+  "runtime": "org.freedesktop.Platform",
+  "runtime-version": "1.6",
+  "sdk": "org.freedesktop.Sdk",
+  "command": "cowsrevenge",
+  "finish-args": [
+     "--socket=x11",
```
Comment (nedrichards) [inline com.github.dariasteam.cowsrevenge.json]: you need "--device=dri" here to get access to the GPU.

**Comment by Dariasteam** [comment]
About the appstream warning, is that an output of running `desktop-file-validate com.github.dariasteam.cowsrevenge.desktop`? because I have none.

**Comment by nedrichards** [comment]
I get the following error when running the app. Doesn't seem to cause any problems though:

> ERROR: load_interactive: Condition ' err != OK ' is true. returned: Ref<ResourceInteractiveLoader>()
   At: core/io/resource_format_binary.cpp:1161.
ERROR: load: Method/Function Failed, returning: RES()
   At: core/io/resource_loader.cpp:210.

**Comment by nedrichards** [comment]
The warning is an output of me running `flatpak-builder --force-clean  --require-changes --repo=repo app com.github.dariasteam.cowsrevenge.json `

**Comment by Dariasteam** [comment]
Yeah, its some obscure godot engine thing I haven't managed to fix but also have never been a problem

**Comment by nedrichards** [comment]
bot, build com.github.dariasteam.cowsrevenge

File: com.github.dariasteam.cowsrevenge.json [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    "split" : false,
+    "no-debuginfo" : true
+  },
+  "modules": [
+    {
+      "name": "cowsrevenge",
+      "sources": [
+        {
+          "type": "git",
+          "url": "https://github.com/Dariasteam/Cows-Revenge-Flatpak",
```
Comment (TingPing) [inline com.github.dariasteam.cowsrevenge.json]: The sources should be either here or upstream, I'm not sure I understand having an intermediate repo. Icons, desktop files, and appdata are not Flatpak specific and would be reused for all packages.

Comment (Dariasteam) [reply com.github.dariasteam.cowsrevenge.json]: The idea of the intermediate repo is to separate the completely free source code and art stuff from the binary. My original intention was to create a makefile that downloads the full godot sources, compiles it and then generates the game itself (similar to what f-droid use) but I had some problems and decided to do this weird thing. Also I wrote the metadata files by the way when reading the flatpak needs so the result is this mess. Should I retry the engine compiling method and use a single repository?

Comment (TingPing) [reply com.github.dariasteam.cowsrevenge.json]: > My original intention was to create a makefile that downloads the full godot sources, compiles it and then generates the game itself 

Yes that is the ideal solution.

Here is an example of building Godot: https://github.com/flathub/org.godotengine.Godot/blob/master/org.godotengine.Godot.json#L34-L72

Comment (Dariasteam) [reply com.github.dariasteam.cowsrevenge.json]: Ok, thank you for the example

Comment (nedrichards) [reply com.github.dariasteam.cowsrevenge.json]: I was wondering if we could import godot as a baseapp and then remove it afterwards, in a manner similar to an SDK - that would make it a bit quicker to build - but that's more idle musing as to an opimization, it wouldn't make a difference to the outcome.

Comment (Dariasteam) [reply com.github.dariasteam.cowsrevenge.json]: It looks like @TingPing is already in it? https://github.com/godotengine/godot/issues/5597

Comment (Dariasteam) [reply com.github.dariasteam.cowsrevenge.json]: Question: Is the build process of flatpak-builder forbidding to acces access internet? This [makefile](https://github.com/Dariasteam/Cows-Revenge/blob/stable/Makefile) successfuly download godot sources when called manually but returns `fatal: unable to access 'https://github.com/godotengine/godot/': Couldn't resolve host 'github.com'
` when called by flatpak-builder

Comment (TingPing) [reply com.github.dariasteam.cowsrevenge.json]: >  Is the build process of flatpak-builder forbidding to acces access internet?

Yes. All sources should be downloaded by `flatpak-builder`.

Comment (Dariasteam) [reply com.github.dariasteam.cowsrevenge.json]: I'm having some problems with this. In order to export godot requires to put some templates files in ~/.local/share/godot and apparently theres no way to specify other path using flags. How should I deal with this considering flatpak is not allowing me to do stuff in ~/ (I have already tried "using --persist=.local/share/godot" but it looks only useful when running not compiling)

Comment (TingPing) [reply com.github.dariasteam.cowsrevenge.json]: > In order to export godot requires to put some templates files in ~/.local/share/godot and apparently theres no way to specify other path using flags.

Honestly the best solution is just to write a patch fixing that in godot.

Comment (Dariasteam) [reply com.github.dariasteam.cowsrevenge.json]: Ok, it can be fixed using [unix_global_settings_path](https://github.com/godotengine/godot-docs/blob/master/development/compiling/packaging_godot.rst), it is also used in your example. Sorry.

**Comment by Dariasteam** [comment]
How about this new super shiny json file?

**Comment by nedrichards** [comment]
bot, build com.github.dariasteam.cowsrevenge

**Comment by nedrichards** [comment]
@Dariasteam Running a test now, superficially looks good at a glance. I know that godot will build on i386 and arm as well in flathub. After this is done it would worth relaxing the flathub.json to do a test build on those platforms

File: com.github.dariasteam.cowsrevenge.json [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+      "sources": [
+        {
+          "type": "git",
+          "url": "https://github.com/godotengine/godot",
+          "commit": "a524cc92e7cccd012d028d870e4648c59473d04e",
+          "branch": "2.1"
+        }
+      ],
+      "build-commands": [
+          "scons platform=server \\
```
Comment (TingPing) [inline com.github.dariasteam.cowsrevenge.json]: Why did you remove the options the other one used? Using host dependencies *should* just be a positive.

File: com.github.dariasteam.cowsrevenge.json [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+        "mv * /app/cowsrevenge/",
+        "cd /app/cowsrevenge/; \\
+          mkdir -p build; \\
+          ../bin/godot-bin -export 'Linux X11' build/cowsrevenge",
+
+        "mkdir -p /app/bin/",
+        "mkdir -p /app/share/applications/",
+        "mkdir -p /app/share/metainfo/",
+        "mkdir -p /app/share/icons/hicolor/64x64/apps/",
+        "mkdir -p /app/share/icons/hicolor/128x128/apps/",
```
Comment (TingPing) [inline com.github.dariasteam.cowsrevenge.json]: You can use `install -D` to install files and make a directory in one call (with -m644 for non-binaries)

File: shared-modules/SDL/SDL-1.2.15.json [hunk 11]
```diff
@@ -0,0 +1,36 @@
+{
```
Comment (TingPing) [inline shared-modules/SDL/SDL-1.2.15.json]: This should be a git submodule.

File: com.github.dariasteam.cowsrevenge.json (hunk 9 above)
Comment (Dariasteam) [reply com.github.dariasteam.cowsrevenge.json]: I was trying to find a bug and decided to use the minimal setup I'm familiar with. I totally forgot to rewrite all flags.

File: shared-modules/SDL/SDL-1.2.15.json (hunk 11 above)
Comment (Dariasteam) [reply shared-modules/SDL/SDL-1.2.15.json]: I'm not familiar with submodules, does it means I should only push the .gitmodules file and not the shared_modules folder?

**Comment by TingPing** [comment]
bot, build com.github.dariasteam.cowsrevenge

File: com.github.dariasteam.cowsrevenge.json [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        "install -D -m644 /app/cowsrevenge/linux_build/com.github.dariasteam.cowsrevenge.appdata.xml \\
+                          /app/share/metainfo/com.github.dariasteam.cowsrevenge.appdata.xml",
+        "install -D -m755 /app/cowsrevenge/linux_build/cowsrevenge_icon_64.png \\
+                          /app/share/icons/hicolor/64x64/apps/com.github.dariasteam.cowsrevenge.png",
+        "install -D -m755 /app/cowsrevenge/linux_build/cowsrevenge_icon_128.png \\
+                          /app/share/icons/hicolor/128x128/apps/com.github.dariasteam.cowsrevenge.png",
+
+        "rm -rf /app/cowsrevenge",
+        "rm -rf /app/bin/godot-bin",
+        "rm -rf /app/templates"
```
Comment (TingPing) [inline com.github.dariasteam.cowsrevenge.json]: I think these better belong in the `cleanup` property of godot and its templates.

**Comment by TingPing** [comment]
In your appstream file `desktop-app` should be `desktop-application` for component type.

**Comment by TingPing** [comment]
`  <launchable type="desktop-id">com.github.dariasteam.cowsrevenge</launchable>`

should be:

`  <launchable type="desktop-id">com.github.dariasteam.cowsrevenge.desktop</launchable>`

**Comment by TingPing** [comment]
```xml
      <description>
        8 Playable levels
      </description>
```

Should be a complete sentence and the should be formatted in paragraphs with `<p>` like above.

**Comment by TingPing** [comment]
`      <caption>XXX: Describe the default screenshot</caption>`

Either remove that or make an actual caption.

**Comment by Dariasteam** [comment]
You two probably are the most patient app reviewer ever. Is it better now?

**Comment by Dariasteam** [comment]
I've added a new picture to the appstream file and now I think its ready to be published. Thank you so much for all your guidance!

**Comment by nedrichards** [comment]
bot, build com.github.dariasteam.cowsrevenge

**Comment by nedrichards** [comment]
This doesn't build properly at the moment - something to do with the way glu is being imported: https://flathub.org/builds/#/builders/2/builds/800

**Comment by nedrichards** [comment]
bot, build com.github.dariasteam.cowsrevenge

**Comment by nedrichards** [comment]
Repository has been created: https://github.com/flathub/com.github.dariasteam.cowsrevenge
